### ISSUE TITLE
[Bug]: addStorage is called once no matter how many editors you make or whether the editor is destroyed, and the state persists

### ISSUE THREAD

**Issue by Nantris** [ueberdosis/tiptap#4664]
### Which packages did you experience the bug in?

core

### What Tiptap version are you using?

2.1.12

### What’s the bug you are facing?

`addStorage` is called only once and never again. The `storage` is not registered anew for new editors, even if the old editor is destroyed. This is a major problem and not documented anywhere. Even studying the code I can't understand how this occurs.

**This is a critical issue for us which makes TipTap entirely unusable in production use cases.**

### What browser are you using?

Chrome

### What did you expect to happen?

`addStorage` is linked to the editor, not to some magical arbitrary object that outlives the object it's attached to.

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

**Comment by jasontlouro** [comment]
Also experiencing this, causing all sorts of weird bugs and states carrying over from one editor to another.

**Comment by Nantris** [comment]
I did manage to work around it but I'm sorry to say I can't recall how we did that and since I can't recall exactly which extension of ours was affected, I'm afraid there's no easy way for me to look up our approach. Sorry I can't be of more help!

**Comment by jasontlouro** [comment]
@Nantris for us it is affecting prosemirror plugins used in tiptap extensions specifically, and workarounds we've tried appear to introduce more problems than they solve.

**Comment by hugh-sun-everlaw** [comment]
We experienced this issue as well... Anyone knows a potential fix here?

**Comment by nperez0111** [comment]
> We experienced this issue as well... Anyone knows a potential fix here?

I think that what is happening here is that extension storage is set on Extension creation (as part of the constructor).

I think for each editor instance if you do a `.extend` call just for that instance it will get around this. I'm not saying that this is the right solution for this or anything but if true, we can move forward & figure out the proper solution

**Comment by hugh-sun-everlaw** [comment]
> > We experienced this issue as well... Anyone knows a potential fix here?
> 
> I think that what is happening here is that extension storage is set on Extension creation (as part of the constructor).
> 
> I think for each editor instance if you do a `.extend` call just for that instance it will get around this. I'm not saying that this is the right solution for this or anything but if true, we can move forward & figure out the proper solution

Hey, thanks for the quick reply here!

What do you mean by this:
> I think for each editor instance if you do a .extend call just for that instance it will get around this.
?

Let's say currently I have code like this:
```typescript
import { Editor, Extension, Extensions } from "@tiptap/core";

const editor1 = new Editor({ extensions: [..., myCustomExtension] });
const editor2 = new Editor({ extensions: [..., myCustomExtension] });
```
And as you said, this has the problem that `myCustomExtension.storage` of `editor1` will refer to the `myCustomExtension.storage` of `editor2`.

Are you saying we should do something like this?
```typescript
import { Editor, Extension, Extensions } from "@tiptap/core";

const editor1 = new Editor({ extensions: [..., myCustomExtension.extend()] });
const editor2 = new Editor({ extensions: [..., myCustomExtension.extend()] });
```
?

Thanks again!

**Comment by nperez0111** [comment]
Yep, I'm saying exactly that (I haven't tested this myself, but if my theory is right that is all it should need).

Unsure whether you need to give it at least an empty object or not. But I think that is the gist of my idea @hugh-sun-everlaw

**Comment by hugh-sun-everlaw** [comment]
> Yep, I'm saying exactly that (I haven't tested this myself, but if my theory is right that is all it should need).
> 
> Unsure whether you need to give it at least an empty object or not. But I think that is the gist of my idea @hugh-sun-everlaw

Gotcha! Thanks! I just tested this idea, and it works like a charm! Before TipTap figures out a solution for this, I am just going to use the solution you suggested [here](https://github.com/ueberdosis/tiptap/issues/4664#issuecomment-2200878320) for now.

Thanks!!!

**Comment by tiagocmendes** [comment]
@nperez0111 I just tested your solution and it worked perfectly! Thank you.

**Comment by nperez0111** [comment]
Related: https://github.com/ueberdosis/tiptap/issues/2694
